doc/specs/#7335 - Console Allocation Policy.md

```diff
@@ -0,0 +1,396 @@
+last updated: 2023-12-12
+issue id: "#7335"
+---
+
+# Console Allocation Policy
+
+## Abstract
+
+Due to the design of the console subsystem on Windows as it has existed since Windows 95, every application that is
+stamped with the `IMAGE_SUBSYSTEM_WINDOWS_CUI` subsystem in its PE header will be allocated a console by kernel32.
+
+Any application that is stamped `IMAGE_SUBSYSTEM_WINDOWS_GUI` will not automatically be allocated a console.
+
+This has worked fine for many years: when you double-click a console application in your GUI shell, it is allocated a
+console.  When you run a GUI application from your console shell, it is **not** allocated a console. The shell will
+**not** wait for it to exit before returning you to a prompt.
+
+There is a large class of applications that do not fit neatly into this mold. Take Python, Ruby, Perl, Lua, or even
+VBScript: These languages are not relegated to running in a console session; they can be used to write fully-fledged GUI
+applications like any other language.
+
+Because their interpreters are console subsystem applications, however, any user double-clicking a shortcut to a Python
+or Perl application will be presented with a console window that the language runtime may choose to garbage collect, or
+may choose not to.
+
+If the runtime chooses to hide the window, there will still be a brief period during which that window is visible. It is
+inescapable.
+
+Likewise, any user running that GUI application from a console shell will see their shell hang until the application
+terminates.
+
+All of these scripting languages worked around this by shipping two binaries each, identical in every way expect in
+their subsystem fields. python/pythonw, perl/perlw, ruby/rubyw, wscript/cscript.
+
+PowerShell[^1] is waiting to deal with this problem because they don't necessarily want to ship a `pwshw.exe` for all
+of their GUI-only authors. Every additional `*w` version of an application is an additional maintenance burden and
+source of cognitive overhead[^2] for users.
+
+On the other side, you have mostly-GUI applications that want to print output to a console **if there is one
+connected**.
+
+These applications are still primarily GUI-driven, but they might support arguments like `/?` or `--help`. They only
+need a console when they need to print out some text. Sometimes they'll allocate their own console (which opens a new
+window) to display in, and sometimes they'll reattach to the originating console. VSCode does the latter, and so when
+you run `code` from CMD, and then `exit` CMD, your console window sticks around because VSCode is still attached to it.
+It will never print anything, and your only option is to close it.
+
+There's another risk in reattaching, too. Given that the shell decides whether to wait based on the subsystem
+field, GUI subsystem applications that reattach to their owning consoles *just to print some text* end up stomping on
+the output of any shell that doesn't wait for them:

```

Comment:
Any chance to address this one as well?

E.g. could a shell guard its console somehow, so that any attempt to `AttachConsole(ATTACH_PARENT_PROCESS)` from a child `IMAGE_SUBSYSTEM_WINDOWS_GUI` process would do nothing and return false, as if there is nothing to attach to?

This behavior is [annoying](https://github.com/FarGroup/FarManager/issues/495), rude, and it happens more and more often.

Reply:
A shell needs to check if the application is GUI/CUI to determine if it should wait or not for the application to finish right? In that case, why not pass `bInheritHandles = FALSE` during `CreateProcess` if it's a CUI application?

Reply:
- That is exactly what we already do
- `AttachConsole` [overrides it anyway](https://learn.microsoft.com/en-us/windows/console/attachconsole):
  > the standard handles retrieved with GetStdHandle will likely be invalid on startup **until AttachConsole is called**. The exception to this is if the application is launched with handle inheritance by its parent process
- `CreateProcess(some_gui_process)` can be called opaquely from `ShellExecuteEx(some_text_file)`

Reply:
Oh, the problem in my test was that my test ""shell"" exited before the GUI application had launched. This caused the `AttachConsole` call to fail.

The way `ATTACH_PARENT_PROCESS` works is that it gets the `InheritedFromUniqueProcessId` value via NT APIs, opens a handle to your process using that PID, and gets the connection HANDLE from your PEB (`NtCurrentTeb()->ProcessEnvironmentBlock->ProcessParameters->Reserved2[0]`).

Locking the console for new attachments is a little difficult, because on one hand the shell doesn't own the console session, and on another, in a way, there's no difference between `AttachConsole` and any other attachments. We could solve these issues of course (e.g. by keeping track which process requested no interference and lifting the restriction once that process is gone), but it does seem non-trivial to me. Would you like me to open an issue so we can track this as a backlog task?

Potential solutions in the meantime:
* Use `PROC_THREAD_ATTRIBUTE_PARENT_PROCESS` and pass a handle of `explorer.exe`. That requires something like `Process32First/NextW` though which is meh.
* Launch a small intermediate process that launches the target GUI application and then immediately exits. This breaks the parent/child relationship and so it can't attach anymore.
* Forbid access to your process (not entirely sure what APIs one could use for that).